AEP/aep-00007.md

```diff
@@ -0,0 +1,225 @@
+| Pulsar        | both     |
+
+## Trade-offs and Constraints
+
+Rewriting the codebase to adopt a fully event-driven, push-based pattern is not feasible given the complexity and scale of such a transformation. The current design prioritizes simplicity and robustness over a pure event-driven model. Considering the long duration of tasks and the cost of orchestration, some inefficiencies during message reception are acceptable. These inefficiencies are preferable to the development overhead and technical debt involved in transforming the existing pull-based mechanism.
+
+The technical limitations of the underlying queueing service also play a role in the design decisions. SQS, for instance, does not provide an API for push-based message reception, which necessitates the use of a pull-based approach. Additionally, past attempts to use RabbitMQ for push-based message handling revealed significant stability issues, such as connection losses and inconsistent message processing. These challenges have further reinforced the decision to rely on the pull-based model, which offers greater reliability and predictability.
+
+We can consider converting a push-based reception to messages into our pull-based interfaces. However, this implementation requires to be as stable as the current one. Pull requests are welcome to improve the implementation of our queue system adaptors.
+
+## Current Interface
+
+Queue interfaces can be found in our dotnet package [ArmoniK.Core](https://www.nuget.org/packages/ArmoniK.Core.Base).
+It contains everything needed to implement queue plugins that can be dynamically loaded by ArmoniK.Core allowing users to implement plugins that match their requirements.
+
+An agent from a partition calls the `PullMessagesAsync` to get messages representing tasks from the associated partition.
+The partition is given outside the `IPullQueueStorage` and should be passed to the implementation through other means.
+It is usually done with .Net options system and environment variables.
+This also means there is no uniformization to set up the partition and it will depend on the implementation of the interface.
+
+```csharp
+/// <summary>
+///   Interface to retrieve messages from the queue
+/// </summary>
+public interface IPullQueueStorage : IQueueStorage
+{
+  /// <summary>
+  ///   Gets messages from the queue
+  /// </summary>
+  /// <param name="nbMessages">Number of messages to retrieve</param>
+  /// <param name="cancellationToken">Token used to cancel the execution of the method</param>
+  /// <returns>
+  ///   Enumerator allowing async iteration over the message queue
+  /// </returns>
+  IAsyncEnumerable<IQueueMessageHandler> PullMessagesAsync(int               nbMessages,
+                                                           CancellationToken cancellationToken = default);
+}
+```
+
+The `PullMessagesAsync` method returns an `IAsyncEnumerable<IQueueMessageHandler>` where the `IQueueMessageHandler` is the following interface.
```

Comment:
I'd write: "where the `IQueueMessageHandler` is an interface representing the life cycle of a message:" 

and move 

The `QueueMessageStatus` is set by ArmoniK during the execution of the task.
Then, the `DisposeAsync` method inherited from `IAsyncDisposable` is used to process the message from the queue by acknowledging it, not acknowledging it, or requeuing it depending on the status.

after the interface definition.